pandas/core/strings.py

```diff
@@ -1962,11 +1935,298 @@ def cons_row(x):
+        # will not need `ignore_index` and the second boolean output anymore
+
+        from pandas import Index, Series, DataFrame, isnull
+
+        # self._orig is either Series or Index
+        idx = self._orig if isinstance(self._orig, Index) else self._orig.index
+
+        err_msg = ('others must be Series, Index, DataFrame, np.ndarrary or '
+                   'list-like (either containing only strings or containing '
+                   'only objects of type Series/Index/list-like/np.ndarray)')
+
+        if isinstance(others, Series):
+            fu_wrn = not others.index.equals(idx)
+            los = [Series(others.values, index=idx)
+                   if ignore_index and fu_wrn else others]
+            return (los, fu_wrn)
+        elif isinstance(others, Index):
+            fu_wrn = not others.equals(idx)
+            los = [Series(others.values,
+                          index=(idx if ignore_index else others))]
+            return (los, fu_wrn)
+        elif isinstance(others, DataFrame):
+            fu_wrn = not others.index.equals(idx)
+            if ignore_index and fu_wrn:
+                # without copy, this could change "others"
+                # that was passed to str.cat
+                others = others.copy()
+                others.index = idx
+            return ([others[x] for x in others], fu_wrn)
+        elif isinstance(others, np.ndarray) and others.ndim == 2:
+            others = DataFrame(others, index=idx)
+            return ([others[x] for x in others], False)
+        elif is_list_like(others):
+            others = list(others)  # ensure iterators do not get read twice etc
+            if all(is_list_like(x) for x in others):
+                los = []
+                fu_wrn = False
+                while others:
+                    nxt = others.pop(0)  # list-like as per check above
+                    # safety for iterators and other non-persistent list-likes
+                    # do not map indexed/typed objects; would lose information
+                    if not isinstance(nxt, (DataFrame, Series,
+                                            Index, np.ndarray)):
+                        nxt = list(nxt)
+
+                    # known types without deep inspection
+                    no_deep = ((isinstance(nxt, np.ndarray) and nxt.ndim == 1)
+                               or isinstance(nxt, (Series, Index)))
+                    # Nested list-likes are forbidden - elements of nxt must be
+                    # strings/NaN/None. Need to robustify NaN-check against
+                    # x in nxt being list-like (otherwise ambiguous boolean)
+                    is_legal = ((no_deep and nxt.dtype == object)
+                                or all((isinstance(x, compat.string_types)
+                                        or (not is_list_like(x) and isnull(x))
+                                        or x is None)
```

Comment:
isnull already checks for None
only 1d objects are valid here (or all scalars)

do this check up front